haystack/preview/components/reader.py

```diff
@@ -0,0 +1,239 @@
+        end = torch.where(mask, end, -torch.inf)  # type: ignore # see above
+        start = start.unsqueeze(-1)
+        end = end.unsqueeze(-2)
+
+        logits = start + end  # shape: (batch_size, seq_length (start), seq_length (end))
+        mask = torch.ones(logits.shape[-2:], dtype=bool)
+        mask = torch.triu(mask)  # End shouldn't be before start
+        mask[0, :1] = False
+        masked_logits = torch.where(mask, logits, -torch.inf)
+        probabilities = self._softmax_across_documents(masked_logits, query_ids)
+
+        flat_probabilities = probabilities.flatten(-2, -1)  # necessary for topk
+        candidates = torch.topk(flat_probabilities, top_k)
+        seq_length = logits.shape[-1]
+        start_candidates = candidates.indices // seq_length  # Recover indices from flattening
+        end_candidates = candidates.indices % seq_length
+        start_candidates = start_candidates.cpu()
+        end_candidates = end_candidates.cpu()
+
+        start_candidates = [
+            [encoding.token_to_chars(start)[0] if start != 0 else 0 for start in candidates]
+            for candidates, encoding in zip(start_candidates, encodings)
+        ]
+        end_candidates = [
+            [encoding.token_to_chars(end)[1] if end != 0 else 0 for end in candidates]
+            for candidates, encoding in zip(end_candidates, encodings)
+        ]
+        probabilities = candidates.values.cpu().tolist()
+
+        return start_candidates, end_candidates, probabilities
+
+    def _unflatten(
+        self,
+        start: List[List[int]],
+        end: List[List[int]],
+        probabilities: List[List[float]],
+        flattened_documents: List[str],
+        documents: List[List[Document]],
+        top_k: int,
+        query_ids: List[int],
+        document_ids: List[int],
+    ) -> List[List[Answer]]:
+        answers = [
+            (flattened_documents[document_id][start:end], probability)
+            for document_id, start_candidates_, end_candidates_, probabilities_ in zip(
+                document_ids, start, end, probabilities
+            )
+            for start, end, probability in zip(start_candidates_, end_candidates_, probabilities_)
+        ]
+
+        i = 0
+        nested_answers = []
+        for query_id in range(query_ids[-1] + 1):
+            current_answers = []
+            while i < len(answers) and query_ids[i // top_k] == query_id:
+                current_answers.append(answers[i])
+                i += 1
+            current_answers = sorted(current_answers, key=lambda answer: answer[1], reverse=True)[:top_k]
+            nested_answers.append(current_answers)
+
+        return nested_answers
+
+    @component.output_types(answers=List[List[Answer]])
+    def run(
+        self,
+        queries: List[str],
+        documents: List[List[Document]],
+        top_k: Optional[int] = None,
+        max_seq_length: Optional[int] = None,
+        stride: Optional[int] = None,
+    ):
```

Comment:
What's the stride? Btw, let's add docstrings so I can find these answers on my own :smile:

Reply:
Ok, added doc strings. When an input is bigger than max_seq_length, we use a sliding window approach and stride is the number of overlapping tokens of two windows.